lib/NodeHelpers.js

```diff
@@ -0,0 +1,106 @@
+// store these matrices for each node in extras._pipeline.flatTransform
+function computeFlatTransformScene(scene, nodes) {
+    var rootNodeIDs = scene.nodes;
+    var scratchMat4 = new Matrix4();
+    var flattenTransformFunction = function(node, parent) {
+        var localTransform = getLocalMatrix4(node, scratchMat4);
+        var parentTransform = Matrix4.IDENTITY;
+        if (defined(parent)) {
+            parentTransform = parent.extras._pipeline.flatTransform;
+        }
+        var flatTransform = new Matrix4();
+        Matrix4.multiply(parentTransform, localTransform, flatTransform);
+        node.extras._pipeline.flatTransform = flatTransform;
+    };
+    for (var index in rootNodeIDs) {
+        if (rootNodeIDs.hasOwnProperty(index)) {
+            var rootNode = nodes[rootNodeIDs[index]];
+            depthFirstTraversal(rootNode, nodes, flattenTransformFunction);
+        }
+    }
+}
+
+// get all the nodes in a scene
+function getAllNodesInScene(scene, nodes) {
+    var rootNodeIDs = scene.nodes;
+    var allSceneNodes = [];
+    var addNodeFunction = function(node) {
+        allSceneNodes.push(node);
+    };
+    for (var index in rootNodeIDs) {
+        if (rootNodeIDs.hasOwnProperty(index)) {
+            var rootNode = nodes[rootNodeIDs[index]];
+            depthFirstTraversal(rootNode, nodes, addNodeFunction);
+        }
+    }
+    return allSceneNodes;
+}
+
+// Perform a stack based iterative depth first traversal on the specified node.
+// Apply functionChildParent to all of its children.
+// functionChildParent functions take this node and the parent as arguments
+function depthFirstTraversal(rootNode, allNodes, functionChildParent) {
+    // process the root node
+    functionChildParent(rootNode);
+
+    // start the traversal with the root node in the queue.
+    var nodeStack = [rootNode];
+    while (nodeStack.length > 0) {
+        var currentNode = nodeStack.pop();
+        // apply the function to all children and push them onto the stack
+        if (!defined(currentNode.children)) {
+            continue;
+        }
+        var childrenIDs = currentNode.children;
+        for (var index in childrenIDs) {
+            if (childrenIDs.hasOwnProperty(index)) {
+                var childID = childrenIDs[index];
+                var childNode = allNodes[childID];
+                functionChildParent(childNode, currentNode);
+                nodeStack.push(childNode);
+            }
+        }
+    }
+}
+
+// gets the local matrix of a node, or computes it from TRS
+function getLocalMatrix4(node, result) {
+    if (!defined(result)) {
+        result = new Matrix4();
+    }
+    // if there is a matrix already, return the matrix
+    if (defined(node.matrix)) {
+        return Matrix4.fromColumnMajorArray(node.matrix, result);
+    }
+    // get it from TRS
+    var translation = Cartesian3.fromArray(defaultValue(node.translation, [0, 0, 0]));
```

Comment:
Does this stage depend on `addDefaults`?  Are we sure that we need to add them here?

If so, they should be file-level scratch variables so we don't allocate on each call.